### ISSUE TITLE
Move low-level utilities into another project

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/canjs#3426]
A.  In order to have us release low-level utilities that aren't strongly associated with the CanJS brand (and less likely for people using other technology to use), we might come up with a project name for these utilities.  This also has the advantage that `CanJS` can represent a "framework" more cleanly and only focus on it's core features.
B.  I argue that because most of these utilities will still depend on other utilities in this ecosystem, we should name them with some sort of prefix or suffix.  For example `x-assign` or `assign-x`.
C.  My suggestion was `assign-4u` b/c it's short and catchy.
D.  We are open to other ideas.  But I'd prefer alternate suggestions instead of criticisms of `4u` at this point.

**Comment by andrejewski** [comment]
Alternative suggestion: no prefix or suffix, put it all in the Bitovi org until they get big enough like Done and Steal.

**Comment by justinbmeyer** [comment]
Proposal: `4u`

Examples:

```
ajax-4u
assign-4u
event-4u
dispatch-4u
```

Identity:

- A bunch of low-to mid-level utilities useful to everyone on any project.

**Comment by James0x57** [comment]
`-aug` short for augmentation, as in: added on top of anything
`ndjson-aug`
`assign-aug`
`event-aug`

**Comment by justinbmeyer** [comment]
@andrejewski 

With that idea, can you address:

- What happens when there are naming conflicts?  I'm guessing `assign` and many others are taken.
- What does "big enough" mean?  Individuals are big enough, the collection is big enough? (We've got probably 40 packages already)
- How do these get documented?
- Why "brand later" instead of immediately?
- We generally try to keep our OS out of Bitovi org because its for client projects and it keeps access rights a bit easier for non-bitovi's to contribute.

**Comment by justinbmeyer** [comment]
Proposal: `JavaScript++`

Examples:

```
ajax-pp
assign-pp
event-pp
dispatch-pp
drag-event-pp
```

Identity:

- A bunch of low-to mid-level utilities useful to everyone on any project.


Pro:

- This could effectively be a rebrand of jQuery++
- `pp` sounds funny


Con:

- `pp` sounds funny

**Comment by matthewp** [comment]
Can you use ++ as the suffix there? `ajax-++` or just `ajax++`?

**Comment by justinbmeyer** [comment]
@matthewp I really doubt we can use `++`.  I know this is something urls normally escape.  I bet it would get super weird.  Though it would be awesome if we could.

**Comment by matthewp** [comment]
SEO could be an issue with that name though.

**Comment by James0x57** [comment]
Proposal: `functionality`, or `for utility`

Examples:
```
ajax-fu
assign-fu
ndjson-fu
drag-event-fu
```

Pro:
* Reads like "f you"

**Comment by phillipskevin** [comment]
I'll throw out `nix` - like... the unix philosophy of doing one thing well.

Sounds better as a prefix I think
`nix-ajax`, `nix-assign`, etc

Cons: 
nix is also a word that basically means "reject" or "nothing"
I don't really like it

**Comment by matthewp** [comment]
I don't have a specific name in mind at the moment, but I wanted to comment in case some else likes this idea too and is more creative than me.

I think it can be a useful exercise to think of the name *last* but instead focus on a couple of other things:

1. What is the *essence* of the brand. In this case it is clearly about low-level utilities as the OP describes.
2. Is there an aesthetic that helps the user *feel* that brand?

So my idea is to base the branding around the idea of a construction site. You see the scaffolding a building, people using heavy equipment to build it. Construction signs etc.

![construction-site](https://user-images.githubusercontent.com/361671/28176330-f8a773fe-67c4-11e7-8214-c44b6e84e904.jpg)

![const_signs_web_opt](https://user-images.githubusercontent.com/361671/28176332-faf09438-67c4-11e7-9689-991030a091f8.jpg)

**Comment by justinbmeyer** [comment]
@matthewp what about `-works`:

```
assign-works
drag-event-works
fire-works //<-- not sure what this does, but I want to find out
```

**Comment by matthewp** [comment]
Yeah, that's not bad. Either `works` or maybe even just `work` without the s. Overall branding could be WorkJS with a website with construction theming.

**Comment by justinbmeyer** [comment]
@andrejewski thanks for elaborating.  I'd like a little more around:

> What happens when there are naming conflicts?

Rename to what?  Anything reasonable that's available?


I agree that we don't need to own everything.  I wish we didn't.  Up until 3.0, we didn't (used jQuery).  Although I think this changed for the following reasons:

1.  We tend to find things don't support all the use cases we need (IE9, Node 4), or they support too much.
2.  We like to expose these things so people don't have to double-up on functionality unnecessarily.  We could fix this w/ documentation, and somehow ensuring that every project uses `object-assign`.  But it's easier to have it centralized.
3.  Relying on other people's code has been painful.  Every release of jQuery (which was our util layer) seemed to cause problems.  We can respond much more quickly within our own code.

For example, `can-util/dom/ajax`'s code is from another OS project.  But it had a bunch of bugs, the maintainer responded too slowly to them, so we brought it in house.  

In the end, I think CanJS's top-level code is easier to maintain because we manage `can-util`.

**Comment by cherifGsoul** [comment]
What about javascriptstd (standard library for javascript)

**Comment by justinbmeyer** [comment]
@andrejewski In my experience with what clients care about, initialization performance is often more important than runtime performance.  Though there's the following considerations:

1.  Usually runtime performance is easier to get right than load-time performance.  After all, network is still slow compared to even mobile processing power.  
2.  The overlap created by multiple `object-assign`'s might be small.  It's `559` minified/gzipped bytes.  However, that could quickly add up and be significant.  Especially as there's lots of packages, and likely lots of other utils that would be needed.
3. Most analytics let us know if a user drops before full initialization ... however, they don't tell us that someone dropped b/c some part of the app was taking too long (so clients might be miss-prioritizing based on faulty data).


Also, while it's useful/nice to target our OS to ourselves, we have to be careful.  Our tastes might differ from those of other developers that are worth targeting products towards.  Many people care about the size of their JS payload. 

This is why I agree with @matthewp 's sentiment that we need to focus on who this is for.  IMO, this is targeting people who want a "medium" complexity toolchain similar to lodash.  Especially one more focused on the DOM and things like task-queues, `ReadableStream` transforms, etc.  

Like lodash, they don't want to spend too much time evaluating every package they consume individually, and judging its quality.  They want to have some baseline expectation that it:

- works in specific browsers/environments
- is tested
- is actively supported

In some ways, I see people being very attracted to a hegemony after the `left-pad` incident.  


Another thought is to split this into two projects:

1. A lodash for the `DOM` side of things.  
2. A lodash for the `JS` side of things built around `can-reflect`.  `can-reflect` is designed to allow lodash like utilities on any decorated data type.  `assign( new Map(), {foo: "bar"})`

**Comment by mikemitchel** [comment]
Springboarding off @matthewp & @justinbmeyer  works idea:

wrx == works (insert road sign logo here)

Similar branding to DoneJS and CanJS - it just Works!
but has the nice 3 letter extension, with an X, but the x is not just for show (ie. worx) but takes the place of 2 letters - code efficiency !

```
assign-wrx
ajax-wrx
fire-wrx (what DOES this do ?!?!)
```

**Comment by frank-dspeed** [comment]
i don't like the idea of splitting anything like that into a extra project and i don't like the bitovi/os private repos idea but if you want to do that i can do nothing about it i think thats a good step backward so that i can't follow up any more public as i could befor but if you want a project and name 

i am strongly for bjs- (Bitovi JS) so every one knows its coded by bitovi and also knows its a js script 

but its up to you if you see any pros about that i see only cons i liked the existing structure it needs only more documentation

**Comment by bmomberger-bitovi** [comment]
Cherif said in https://github.com/canjs/canjs/issues/3426#issuecomment-315143438

> What about javascriptstd (standard library for javascript)

I like the idea of promoting it as a standard, but I'd use -sb (for JS Standard Base) or -sp for (JS Starting Point) instead.  Some brief research says that the two letter forms of this aren't used for anything in the JS world, and at very least JSSB only name clashes with a bank.

**Comment by cherifGsoul** [comment]
@bmomberger-bitovi yes that was the idea, I like Standard Base also, I just found https://github.com/stdlib-js/stdlib so this can clashes also with std I want to say bedrock but it is taken also https://github.com/digitalbazaar/bedrock so maybe Javascript Standard Foundation (JSSF) and I found this close to that name.

**Comment by frank-dspeed** [comment]
i think maybe that should get even more splitted up by its tool type for example if its a class or decorator or cjs script something like that
